### ISSUE TITLE
Get Gradebook Grades rounding issue on Input Final Grades screen

### ISSUE THREAD

**Issue by hilltopper06** [francoisjacquet/rosariosis#177]
When you press "Get Gradebook Grades" the proper percentage grades appear in the Letter Percent fields.  When you press the save button, the grades change to end in 5.  For example, 78.2% becomes 75%.  99.9% becomes 95%.  90.1% becomes 95%.  It is not rounding, it keeps the 10's place value and changes the one's place to 5 and removes the decimal point.  This is on the most recent release 2.9.14

**Comment by francoisjacquet** [comment]
Hello @hilltopper06 
Thank you for reporting your issue.

We were not able to reproduce it unfortunately:
![screenshot from 2017-01-05 10 29 40](https://cloud.githubusercontent.com/assets/4723501/21686157/61187258-d332-11e6-89ac-0d50a03155b8.png)

![image](https://cloud.githubusercontent.com/assets/4723501/21686145/56116d88-d332-11e6-916c-2a8b25ce5dc4.png)

Please share your configuration (server / browser), including the RosarioSIS configuration options related to Grades (scales, using letter grades only?... etc) along with the exact steps to reproduce the problem.
Please also attach screenshots.

**Comment by hilltopper06** [comment]
School Configuration as it relates to grades
![actconfig](https://cloud.githubusercontent.com/assets/21183376/21692677/95efbe9e-d343-11e6-93dc-3ae7849863e6.PNG)
Grading Scale used
![gradescale](https://cloud.githubusercontent.com/assets/21183376/21692679/99a3c13e-d343-11e6-8ca5-ef200be9e1cb.PNG)
Clicking the get gradebook grades works!
![getgrades](https://cloud.githubusercontent.com/assets/21183376/21692680/9b4c2274-d343-11e6-8c66-242ce896f51a.PNG)
But once you click save the percentages change as described in my initial post.
![aftersave](https://cloud.githubusercontent.com/assets/21183376/21692684/9e48860c-d343-11e6-89e9-d947020bb39c.PNG)

**Comment by hilltopper06** [comment]
Server is Windows 10 based and follows the instructions from here https://github.com/francoisjacquet/rosariosis/wiki/How-to-install-RosarioSIS-on-Windows

Browser is Version 55.0.2883.87 m (64-bit) of Chrome (although other end users may be using recent versions Firefox or Safari, the issue persists across all browsers tested).

**Comment by francoisjacquet** [comment]
Thank you for the details.

We could not reproduce the issue using Windows 10 and Chrome and your configuration.

Please try to reproduce the errors on the [demo site](https://rosariosis.org/demo/) and share the relevant steps here.

**Comment by hilltopper06** [comment]
I was unable to recreate the issue on the demo site, but I do have some addition information to add.

As mentioned previously, pressing the "get gradebook grades" button populates the final percent grade fields accurately, but then pressing the save button causes them to change.  At first I thought they were all changing to be the same number in the 10s place and 5 in the ones place (92 became 95, 97, became 95, 81 became 85, etc.).  I then noticed that a grade below 60 (53 for instance) became 30.  This made me realize that the results that were being saved to the database were correlated to my cutoff percents.  A's became 95 ((100+90)/2), all the way down to F ((60+0)/2).  I poked around in the function code and noticed this bit in the _makePercentGrade.fnc.php file:

```
foreach ( (array) $_ROSARIO['_makeLetterGrade']['grades'][ $grade_scale_id ] as $grade )
	{
		$prev = $crnt;

		$crnt = ( $does_breakoff === 'Y' ?
			$gradebook_config[ $course_period_id . '-' . $grade['ID'] ] :
			$grade['BREAK_OFF'] );

		if ( is_numeric( $grade_id_or_title ) ?
				$grade_id_or_title == $grade['ID'] :
				mb_strtoupper( $grade_id_or_title ) == mb_strtoupper( $grade['TITLE'] ) )
		{
			return ( $crnt + ( $crnt > $prev ? 100 : $prev ) ) / 2;
		}
	}
```

Is there any chance the the return ( $crnt + ( $crnt > $prev ? 100 : $prev ) ) / 2; line could be effecting my percent grades?

**Comment by francoisjacquet** [comment]
Hello @hilltopper06 

If you want to debug the code used to save the percentage, you could add an echo:

```php
if ( is_numeric( $grade_id_or_title ) ?
		$grade_id_or_title == $grade['ID'] :
		mb_strtoupper( $grade_id_or_title ) == mb_strtoupper( $grade['TITLE'] ) )
{
	echo 'Here'; // Will this be printed when you hit Save in Input Final Grades?
	return ( $crnt + ( $crnt > $prev ? 100 : $prev ) ) / 2;
}
```

**Comment by francoisjacquet** [comment]
Wanted to add the behaviour you describe is just _strange_:

If your Percent grades are **re**calculated when saving, using the `_makePercentGrade()` function (can you confirm it, @hilltopper06 ?), there is a problem.

Indeed, when saving grades, the percent grades are calculated **only if** the percent fields are empty, see https://github.com/francoisjacquet/rosariosis/blob/mobile/modules/Grades/InputFinalGrades.php#L369 and https://github.com/francoisjacquet/rosariosis/blob/mobile/modules/Grades/InputFinalGrades.php#L469

Your screenshots show filled in percent fields.

**Comment by hilltopper06** [comment]
I have been wanting to set up a test environment to continue to troubleshoot the issue (so as not to alter any finalized grades, as they were corrected manually).  Once I am able to I will revisit this issue.  To work around it I used 2 displays, both on the same final grades page.  I would get the grade on one, and instead of pressing save, would manually type in the percent given into the other for each student and then click save on the manual entry page.  This resulted in the proper values being saved.  Luckily there are just under 50 students in my organization, so manual entry was not too tedious.

**Comment by francoisjacquet** [comment]
I can test your database if you will. Please send me a SQL dump / backup by email.

**Comment by lkozloff** [comment]
I'm experiencing this same issue presently, and can reproduce it on our staging server. Happy to share credentials or otherwise contribute to getting this solved.

*edit*
I've gathered a bit more information, without having dived in to the code at all
- if the school is set to **Use Percent Grades Only**: 
    - there is no 'rounding'
    - there is one entry for each grade in `student_report_card_grades`
- if the school is set to **Use Letter and Percent Grades** 
    - there is the same rounding as described above.
    - there are two database entries in `student_report_card_grades`, neither of which is the original score entered
- **Clear all** clears out the scores, but after hitting save scores re-appear. 

I can't reproduce any of this on the demo server.. but can on our staging server.

**Comment by francoisjacquet** [comment]
Hello @lkozloff

Thank you for reporting the same issue about Final Grades input and your will to help us reproduce the error.

Please contact us directly by mail or using the form at https://www.rosariosis.org/services/

François

**Comment by francoisjacquet** [comment]
Hello @hilltopper06 and @lkozloff 

The issue has been fixed. The fix will come with the next release (v3.5.3)

Thank you for your collaboration.

**Comment by lkozloff** [comment]
I pulled the latest version and can confirm the issue appears to be resolved. Thanks so much! Our upcoming marking period will be much easier on me ;)
